Documentation/wip/testing-proposal.md

```diff
@@ -0,0 +1,66 @@
+## Upgrade tests
+
+Currently in Fission, there is no way to run tests between upgrades and check if the upgrade broke something. This proposal aims to take a shot at this and while we are at it, a small diversion into related topic of testing! Please share your feedback and comments.
+
+### Goals
+- Newer version of Fission CLI should work with older version of Fission
+- A newer version of Fission passes all tests from previous version of Fission (Regression)
+- A cluster when upgraded from previous release behaves in the same way as when a newer version is installed from scratch on a cluster
+
+### Non-goals
+- Tests from newer version of Fission working with older versions of Fission
+
+### Proposal
+
+The upgrade test is run only once in a certain frequency (Mostly once a day) based on a cron job. Two version of fission source code are checked out in two subdirectories, the parameters for the version could be specified as environment variables.
+
+```
+./fission
+./fission_old
+```
+
+1. Fission_old is built and deployed in the cluster and then tests are run. This is exact same as the current build and test workflow but for a previous version of Fission.
+
+2. Once (1) passes, script switches to fission directory and builds and pushes the images. The part where the workflow diverts is install_and_test, and here are the differences:
+
+- Instead of ```helm install```, we will run ```helm upgrade``` - and upgrade the existing cluster with newer images
+- Script will use the new fission CLI
+- Script will use tests in fission_old to run the tests
+
+3. The script for (1) can be used from existing codebase while the part (2) will need developing a part of script specifically for upgrade.
+
+4. After the run - if both set of tests succeed, the upgrade can be considered fairly successful (Although not entirely - as there can be potential gap)
+
+### Alternative approach considered
+
+It is possible to get the binary from previous release instead of source code - but it has two drawbacks:
+
+1. The effort for (2) in previous section still remains the same
+2. With binary install, we loose the tests from older version of Fission.
```

Comment:
I don't think I agree with this? The tests are fully separate from installation.

Reply:
What I meant here is if we don't check out the older version of Fission - and only use binary, then the tests from the previous version will have to be retrieved separately, as binary does not container tests. And we need tests from the previous version for "before upgrade" tests.

Reply:
Oh, interesting.  How about using the tests from the current version, with just a header comment?  So just like we have "disabled:true" available right now, we could have a "pre-upgrade: true", and tests with that comment would be run.

Also, note that you'd need tests that don't clean up, so that we can test that functions etc. created before upgrade still work after the upgrade.

Reply:
You make a great point about keeping objects through the upgrade process and then run tests. I think these can be special tests intended for upgrade only and for these tests we can use the tag like "pre-upgrade: true"

The other aspect - which tests not to run before the upgrade, for ex. tests added in 0.7 should not be run at 0.6 level. I don't like the idea of using tags for this purpose. Are we on same page with this or else I can explain a little more.